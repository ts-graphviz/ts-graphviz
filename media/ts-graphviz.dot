digraph {
  rankdir = "LR";
  node [
    shape = "circle";
  ];
  "factory_fn" [
    shape = "plain";
    label = <<table><tr><td bgcolor="#f0f0f0" align="CENTER">Model Factory Functions</td></tr><tr><td port="description" align="LEFT">strict.digraph(...)<br align="LEFT"></br>strict.graph(...)<br align="LEFT"></br>digraph(...)<br align="LEFT"></br>graph(...)<br align="LEFT"></br></td></tr></table>>;
  ];
  "oop" [
    shape = "plain";
    label = <<table><tr><td bgcolor="#f0f0f0" align="CENTER">Object-Oriented Programming</td></tr><tr><td port="description" align="LEFT">new Digraph(...)<br align="LEFT"></br>new Graph(...)<br align="LEFT"></br>...<br align="LEFT"></br></td></tr></table>>;
  ];
  {"factory_fn" "oop"} -> "Model";
  "Model" -> "DOT" [
    label = "toDot";
  ];
  "DOT" -> "Model" [
    label = "fromDot";
  ];
}
