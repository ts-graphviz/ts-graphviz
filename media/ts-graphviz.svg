<svg xmlns="http://www.w3.org/2000/svg" width="512pt" height="202pt" viewBox="0 0 511.69 201.5"><g class="graph" transform="translate(4 197.5)"><path fill="#fff" stroke="transparent" d="M-4 4v-201.5h511.69V4z"/><g class="node"><path fill="#f0f0f0" stroke="transparent" d="M22.5-169v-21h184v21z"/><path fill="none" stroke="#000" d="M22.5-169v-21h184v21z"/><text x="25.5" y="-175.8" font-family="Times,serif" font-size="14">Model Factory Functions</text><path fill="none" stroke="#000" d="M22.5-101v-66h184v66z"/><text x="25.5" y="-152.8" font-family="Times,serif" font-size="14">strict.digraph(...)</text><text x="25.5" y="-137.8" font-family="Times,serif" font-size="14">strict.graph(...)</text><text x="25.5" y="-122.8" font-family="Times,serif" font-size="14">digraph(...)</text><text x="25.5" y="-107.8" font-family="Times,serif" font-size="14">graph(...)</text><path fill="none" stroke="#000" d="M19.5-98.5v-95h190v95z"/></g><g class="node"><circle cx="305.65" cy="-93" r="39.79" fill="none" stroke="#000"/><text x="305.65" y="-89.3" font-family="Times,serif" font-size="14" text-anchor="middle">Model</text></g><g class="edge"><path fill="none" stroke="#000" d="M209.52-119.66c16.57 4.65 33.17 9.3 47.75 13.38"/><path stroke="#000" d="m258.33-109.61 8.69 6.07-10.57.67z"/></g><g class="node"><path fill="#f0f0f0" stroke="transparent" d="M3.5-56v-21h223v21z"/><path fill="none" stroke="#000" d="M3.5-56v-21h223v21z"/><text x="6.5" y="-62.8" font-family="Times,serif" font-size="14">Object-Oriented Programming</text><path fill="none" stroke="#000" d="M3.5-3v-51h223v51z"/><text x="6.5" y="-39.8" font-family="Times,serif" font-size="14">new Digraph(...)</text><text x="6.5" y="-24.8" font-family="Times,serif" font-size="14">new Graph(...)</text><text x="6.5" y="-9.8" font-family="Times,serif" font-size="14">...</text><path fill="none" stroke="#000" d="M0 0v-80h229V0z"/></g><g class="edge"><path fill="none" stroke="#000" d="m229.08-71.82 28.42-7.97"/><path stroke="#000" d="m256.6-83.17 10.58.67-8.69 6.07z"/></g><g class="node"><circle cx="472.49" cy="-93" r="31.4" fill="none" stroke="#000"/><text x="472.49" y="-89.3" font-family="Times,serif" font-size="14" text-anchor="middle">DOT</text></g><g class="edge"><path fill="none" stroke="#000" d="M345.3-93h85.43"/><path stroke="#000" d="m431.05-96.5 10 3.5-10 3.5z"/><text x="393.29" y="-96.8" font-family="Times,serif" font-size="14" text-anchor="middle">toDot</text></g><g class="edge"><path fill="none" stroke="#000" d="M443.99-80.12c-6.64 2.54-13.8 4.81-20.7 6.12-26.2 4.98-33.68 4.3-60 0-3.48-.57-7.04-1.32-10.59-2.2"/><path stroke="#000" d="m351.47-72.91-8.7-6.05 10.58-.69z"/><text x="393.29" y="-77.8" font-family="Times,serif" font-size="14" text-anchor="middle">fromDot</text></g></g></svg>