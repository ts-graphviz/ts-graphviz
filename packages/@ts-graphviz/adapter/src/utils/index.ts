export { createCommandAndArgs } from './create-command-and-args.js';
