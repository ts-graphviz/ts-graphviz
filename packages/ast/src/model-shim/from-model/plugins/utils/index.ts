export * from './convert-attribute.js';
export * from './convert-cluster-children.js';
export * from './convert-comment.js';
