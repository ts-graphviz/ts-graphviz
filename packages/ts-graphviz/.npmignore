jest.config.json
README_ja.md
rollup.config.js
src
test
tsconfig.build.json
tsconfig.json
.turbo
