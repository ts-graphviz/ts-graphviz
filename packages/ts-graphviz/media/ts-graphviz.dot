digraph state_machine {
  rankdir = LR;
  node [ shape = circle ]

  // { rank = same; factory_fn; Model; DOT; }

  factory_fn [
    shape = record,
    label = "{{Model Factory Functions | strict.digraph(...)\lstrict.graph(...)\ldigraph(...)\lgraph(...)\l}}"
  ]

  oop [
    shape = record,
    label = "{{Object Oriented Programing | new Digraph(...)\lnew Graph(...)\l...\l}}"
  ]

  {factory_fn, oop} -> Model;

  Model -> DOT [ label = "toDot" ]
  DOT -> Model [ label = "fromDot" ]
}
