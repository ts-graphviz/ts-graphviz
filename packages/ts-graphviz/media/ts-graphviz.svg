<svg width="503pt" height="196pt" viewBox="0 0 502.69 195.5" xmlns="http://www.w3.org/2000/svg"><g class="graph" transform="translate(4 191.5)"><path fill="#fff" stroke="transparent" d="M-4 4v-195.5h502.69V4H-4z"/><g class="node"><path fill="none" stroke="#000" d="M13-96v-91h194v91H13z"/><text text-anchor="middle" x="110" y="-171.8" font-family="Times,serif" font-size="14">Model Factory Functions</text><path fill="none" stroke="#000" d="M13-164h194"/><text x="21" y="-148.8" font-family="Times,serif" font-size="14">strict.digraph(...)</text><text x="21" y="-133.8" font-family="Times,serif" font-size="14">strict.graph(...)</text><text x="21" y="-118.8" font-family="Times,serif" font-size="14">digraph(...)</text><text x="21" y="-103.8" font-family="Times,serif" font-size="14">graph(...)</text></g><g class="node"><circle fill="none" stroke="#000" cx="296.65" cy="-89.5" r="39.79"/><text text-anchor="middle" x="296.65" y="-85.8" font-family="Times,serif" font-size="14">Model</text></g><g class="edge"><path fill="none" stroke="#000" d="M207.11-114.43c14.38 4.05 28.64 8.06 41.38 11.65"/><path stroke="#000" d="m249.67-106.08 8.68 6.08-10.57.65 1.89-6.73z"/></g><g class="node"><path fill="none" stroke="#000" d="M0-.5v-76h220v76H0z"/><text text-anchor="middle" x="110" y="-61.3" font-family="Times,serif" font-size="14">Object Oriented Programing</text><path fill="none" stroke="#000" d="M0-53.5h220"/><text x="8" y="-38.3" font-family="Times,serif" font-size="14">new Digraph(...)</text><text x="8" y="-23.3" font-family="Times,serif" font-size="14">new Graph(...)</text><text x="8" y="-8.3" font-family="Times,serif" font-size="14">...</text></g><g class="edge"><path fill="none" stroke="#000" d="M220.01-68.61c9.87-2.73 19.48-5.38 28.35-7.83"/><path stroke="#000" d="m247.47-79.82 10.57.71-8.7 6.03-1.87-6.74z"/></g><g class="node"><circle fill="none" stroke="#000" cx="463.49" cy="-89.5" r="31.4"/><text text-anchor="middle" x="463.49" y="-85.8" font-family="Times,serif" font-size="14">DOT</text></g><g class="edge"><path fill="none" stroke="#000" d="M336.3-89.5h85.43"/><path stroke="#000" d="m422.05-93 10 3.5-10 3.5v-7z"/><text text-anchor="middle" x="384.29" y="-93.3" font-family="Times,serif" font-size="14">toDot</text></g><g class="edge"><path fill="none" stroke="#000" d="M434.99-76.62c-6.64 2.54-13.8 4.81-20.7 6.12-26.2 4.98-33.68 4.3-60 0-3.48-.57-7.04-1.32-10.59-2.2"/><path stroke="#000" d="m342.47-69.41-8.7-6.05 10.58-.69-1.88 6.74z"/><text text-anchor="middle" x="384.29" y="-74.3" font-family="Times,serif" font-size="14">fromDot</text></g></g></svg>
