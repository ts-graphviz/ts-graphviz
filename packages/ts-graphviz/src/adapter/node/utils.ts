import { pipeline as _pipeline } from 'node:stream';
import { promisify } from 'node:util';

/**
 * NOTE:
 * The node:stream/promises standard module is not provided in Node 14.
 * Fix Node 14 to use node:stream/promises after LTS ends.
 */
export const pipeline = promisify(_pipeline);
