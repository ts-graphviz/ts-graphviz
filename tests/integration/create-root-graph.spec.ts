import { digraph, graph, strict, Graph, Digraph, attribute as _ } from 'ts-graphviz';
import { toDot } from '#test/utils';

describe('function digraph', () => {
  it('should return Graph object, when execute digraph()', () => {
    const g = digraph();
    expect(g).toBeInstanceOf(Digraph);
    expect(g.strict).toBe(false);
    expect(g.directed).toBe(true);
  });

  describe('root create function', () => {
    test.each([
      ['id', { size: 0, id: 'id', strictMode: false, g: digraph('id') }],
      [
        'id with attributes',
        {
          size: 1,
          id: 'id',
          strictMode: false,
          g: digraph('id', {
            [_.label]: 'Label',
          }),
        },
      ],
      [
        'no parameters',
        {
          size: 0,
          id: undefined,
          strictMode: false,
          g: digraph(),
        },
      ],
      [
        'strict no parameters',
        {
          size: 0,
          id: undefined,
          strictMode: true,
          g: strict.digraph(),
        },
      ],
      [
        'no id with attributes',
        {
          size: 1,
          id: undefined,
          strictMode: false,
          g: digraph({
            [_.label]: 'Label',
          }),
        },
      ],
      [
        'strict with attributes',
        {
          size: 1,
          id: undefined,
          strictMode: true,
          g: strict.digraph({
            [_.label]: 'Label',
          }),
        },
      ],
      [
        'strict id with attributes',
        {
          size: 1,
          id: 'id',
          strictMode: true,
          g: strict.digraph('id', {
            [_.label]: 'Label',
          }),
        },
      ],
      [
        'strict no id with attributes',
        {
          size: 1,
          id: undefined,
          strictMode: true,
          g: strict.digraph({
            [_.label]: 'Label',
          }),
        },
      ],
    ])('%s', (_, { id, size, strictMode, g }) => {
      expect(g.id).toBe(id);
      expect(g.size).toBe(size);
      expect(g.strict).toBe(strictMode);
      expect(toDot(g)).toMatchSnapshot();
    });
  });

  test('callback style', () => {
    const G = digraph('G', (g) => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], (e) => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', (A) => {
        const Aa = A.node('Aaa', (n) => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', (n) => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa.port('a'), Ab, Ac, 'E'], (e) => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });

  test('callback style, set attributes by attributes object', () => {
    const G = digraph('G', (g) => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], {
        [_.color]: 'red',
      });
      g.subgraph('A', (A) => {
        const Aa = A.node('Aaa', {
          [_.color]: 'pink',
        });

        const Ab = A.node('Abb', {
          [_.color]: 'violet',
        });
        const Ac = A.node('Acc');
        A.edge([Aa.port('a'), Ab, Ac, 'E'], {
          [_.color]: 'red',
        });
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });

  test('comment', () => {
    const G = digraph('G', (g) => {
      g.comment = 'This is directed graph.';
      const a = g.node('aa');
      a.comment = 'This is node a.';
      const b = g.node('bb');
      b.comment = 'This is node b.';
      const c = g.node('cc');
      c.comment = 'This is node c.';
      g.edge([a, b, c], (e) => {
        e.comment = 'This is edge.\nIt connects a, b and c.';

        e.attributes.set('color', 'red');
        e.attributes.comment = 'Edge line will draw with red.';
      });
      g.subgraph('A', (A) => {
        A.comment = 'It is subgraph A.\nIt is not cluster';
        A.node('Aaa', (n) => {
          n.comment = 'This is node Aaa in subgraph A.';
          n.attributes.comment = 'It will be filled by pink.';
          n.attributes.set('color', 'pink');
        });
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });
});

describe('function graph', () => {
  it('should return Graph object, when execute graph()', () => {
    const g = graph();
    expect(g).toBeInstanceOf(Graph);
  });

  it('script style', () => {
    const g = digraph('G');
    const subgraphA = g.createSubgraph('A');
    const nodeA1 = subgraphA.createNode('A_node1');
    const nodeA2 = subgraphA.createNode('A_node2');
    subgraphA.createEdge([nodeA1, nodeA2]);

    const subgraphB = g.createSubgraph('B');
    const nodeB1 = subgraphB.createNode('B_node1');
    const nodeB2 = subgraphB.createNode('B_node2');
    subgraphA.createEdge([nodeB1, nodeB2]);

    const node1 = g.createNode('node1');
    const node2 = g.createNode('node2');
    g.createEdge([node1, node2]);
    const dot = toDot(g);
    expect(dot).toMatchSnapshot();
  });

  test('callback style', () => {
    const G = graph('G', (g) => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], (e) => {
        e.attributes.set('color', 'red');
      });
      g.subgraph('A', (A) => {
        const Aa = A.node('Aaa', (n) => {
          n.attributes.set('color', 'pink');
        });
        const Ab = A.node('Abb', (n) => {
          n.attributes.set('color', 'violet');
        });
        const Ac = A.node('Acc');
        A.edge([Aa, Ab, Ac, A.node('hoge').port('fuga')], (e) => {
          e.attributes.set('color', 'red');
        });

        A.edge([Aa.port({ port: 'a', compass: 'w' }), Ab.port({ compass: 'w' }), 'Aaa:e', 'Acc:r:e'], (e) => {
          e.attributes.set('color', 'red');
        });
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });

  test('callback style, set attributes by attributes object', () => {
    const G = graph('G', (g) => {
      const a = g.node('aa');
      const b = g.node('bb');
      const c = g.node('cc');
      g.edge([a, b, c], {
        [_.color]: 'red',
      });
      g.subgraph('A', (A) => {
        const Aa = A.node('Aaa', {
          [_.color]: 'pink',
        });

        const Ab = A.node('Abb', {
          [_.color]: 'violet',
        });
        const Ac = A.node('Acc');
        A.edge([Aa.port('a'), Ab, Ac, 'E'], {
          [_.color]: 'red',
        });
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });

  test('escape characters', () => {
    const G = graph('G\n"', (g) => {
      const a = g.node('a\na');
      a.attributes.set('label', '1\n2\n"\n3');
      const b = g.node('b"b');
      const c = g.node('c\nc"');
      g.edge([a, b, c], (e) => {
        e.attributes.set('color', 'red');
      });

      g.subgraph('graph.name', (s) => {
        const innerA = s.node('node.name');
        innerA.attributes.set('label', 'node');
        const innerB = s.node('another.name');
        innerB.attributes.set('label', 'words with space and "quote"');
        s.edge([innerA, innerB]);
      });
    });
    const dot = toDot(G);
    expect(dot).toMatchSnapshot();
  });
});

describe('strict mode', () => {
  it('should return Digraph object, when execute digraph()', () => {
    const g = strict.digraph();
    expect(g).toBeInstanceOf(Digraph);
    expect(g.strict).toBe(true);
    expect(g.directed).toBe(true);
  });

  it('should return Graph object, when execute graph()', () => {
    const g = strict.graph();
    expect(g).toBeInstanceOf(Graph);
    expect(g.strict).toBe(true);
  });
});
