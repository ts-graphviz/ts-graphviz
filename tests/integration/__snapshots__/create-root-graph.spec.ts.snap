// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`function digraph callback style 1`] = `
digraph "G" {
  "aa";
  "bb";
  "cc";
  subgraph "A" {
    "Aaa" [
      color = "pink";
    ];
    "Abb" [
      color = "violet";
    ];
    "Acc";
    "Aaa":"a" -> "Abb" -> "Acc" -> "E" [
      color = "red";
    ];
  }
  "aa" -> "bb" -> "cc" [
    color = "red";
  ];
}
`;

exports[`function digraph callback style, set attributes by attributes object 1`] = `
digraph "G" {
  "aa";
  "bb";
  "cc";
  subgraph "A" {
    "Aaa" [
      color = "pink";
    ];
    "Abb" [
      color = "violet";
    ];
    "Acc";
    "Aaa":"a" -> "Abb" -> "Acc" -> "E" [
      color = "red";
    ];
  }
  "aa" -> "bb" -> "cc" [
    color = "red";
  ];
}
`;

exports[`function digraph comment 1`] = `
// This is directed graph.
digraph "G" {
  // This is node a.
  "aa";
  // This is node b.
  "bb";
  // This is node c.
  "cc";
  // It is subgraph A.
  // It is not cluster
  subgraph "A" {
    // This is node Aaa in subgraph A.
    "Aaa" [
      // It will be filled by pink.
      color = "pink";
    ];
  }
  // This is edge.
  // It connects a, b and c.
  "aa" -> "bb" -> "cc" [
    // Edge line will draw with red.
    color = "red";
  ];
}
`;

exports[`function digraph root create function id 1`] = `digraph "id" {}`;

exports[`function digraph root create function id with attributes 1`] = `
digraph "id" {
  label = "Label";
}
`;

exports[`function digraph root create function no id with attributes 1`] = `
digraph {
  label = "Label";
}
`;

exports[`function digraph root create function no parameters 1`] = `digraph {}`;

exports[`function digraph root create function strict id with attributes 1`] = `
strict digraph "id" {
  label = "Label";
}
`;

exports[`function digraph root create function strict no id with attributes 1`] = `
strict digraph {
  label = "Label";
}
`;

exports[`function digraph root create function strict no parameters 1`] = `strict digraph {}`;

exports[`function digraph root create function strict with attributes 1`] = `
strict digraph {
  label = "Label";
}
`;

exports[`function graph callback style 1`] = `
graph "G" {
  "aa";
  "bb";
  "cc";
  subgraph "A" {
    "Aaa" [
      color = "pink";
    ];
    "Abb" [
      color = "violet";
    ];
    "Acc";
    "hoge";
    "Aaa" -- "Abb" -- "Acc" -- "hoge":"fuga" [
      color = "red";
    ];
    "Aaa":"a":"w" -- "Abb":"w" -- "Aaa":"e" -- "Acc":"r":"e" [
      color = "red";
    ];
  }
  "aa" -- "bb" -- "cc" [
    color = "red";
  ];
}
`;

exports[`function graph callback style, set attributes by attributes object 1`] = `
graph "G" {
  "aa";
  "bb";
  "cc";
  subgraph "A" {
    "Aaa" [
      color = "pink";
    ];
    "Abb" [
      color = "violet";
    ];
    "Acc";
    "Aaa":"a" -- "Abb" -- "Acc" -- "E" [
      color = "red";
    ];
  }
  "aa" -- "bb" -- "cc" [
    color = "red";
  ];
}
`;

exports[`function graph escape characters 1`] = `
graph "G\\n\\"" {
  "a\\na" [
    label = "1\\n2\\n\\"\\n3";
  ];
  "b\\"b";
  "c\\nc\\"";
  subgraph "graph.name" {
    "node.name" [
      label = "node";
    ];
    "another.name" [
      label = "words with space and \\"quote\\"";
    ];
    "node.name" -- "another.name";
  }
  "a\\na" -- "b\\"b" -- "c\\nc\\"" [
    color = "red";
  ];
}
`;

exports[`function graph script style 1`] = `
digraph "G" {
  "node1";
  "node2";
  subgraph "A" {
    "A_node1";
    "A_node2";
    "A_node1" -> "A_node2";
    "B_node1" -> "B_node2";
  }
  subgraph "B" {
    "B_node1";
    "B_node2";
  }
  "node1" -> "node2";
}
`;
